From a36d92b618dafd2a3261431ede96416d2183f31e Mon Sep 17 00:00:00 2001
From: Hannes Frederic Sowa <hannes@redhat.com>
Date: Wed, 18 Jun 2014 15:11:04 -0400
Subject: [net] neigh: set nud_state to NUD_INCOMPLETE when probing router reachability

Message-id: <5a5d5bed049b0902305aa4f19d6a447a8cc28fcd.1403103924.git.hannes@redhat.com>
Patchwork-id: 84258
O-Subject: [PATCH RHEL7.1 net] neigh: set nud_state to NUD_INCOMPLETE when probing router reachability
Bugzilla: 1110193
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jesper Brouer <jbrouer@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110193
Upstream Status: 2176d5d41891753774f648b67470398a5acab584
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=7591273

No conflicts.

During the backport of BZ#1029585 to RHEL5 this patch got added which
is missing from the RHEL7 series.

Upstream Commit:
commit 2176d5d41891753774f648b67470398a5acab584
Author: Duan Jiong <duanj.fnst@cn.fujitsu.com>
Date:   Fri May 9 13:16:48 2014 +0800

    neigh: set nud_state to NUD_INCOMPLETE when probing router reachability

    Since commit 7e98056964("ipv6: router reachability probing"), a router falls
    into NUD_FAILED will be probed.

    Now if function rt6_select() selects a router which neighbour state is NUD_FAILED,
    and at the same time function rt6_probe() changes the neighbour state to NUD_PROBE,
    then function dst_neigh_output() can directly send packets, but actually the
    neighbour still is unreachable. If we set nud_state to NUD_INCOMPLETE instead
    NUD_PROBE, packets will not be sent out until the neihbour is reachable.

    In addition, because the route should be probes with a single NS, so we must
    set neigh->probes to neigh_max_probes(), then the neigh timer timeout and function
    neigh_timer_handler() will not send other NS Messages.

    Signed-off-by: Duan Jiong <duanj.fnst@cn.fujitsu.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Hannes Frederic Sowa <hannes@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/neighbour.c b/net/core/neighbour.c
index 6bcf819..ed028a2 100644
--- a/net/core/neighbour.c
+++ b/net/core/neighbour.c
@@ -1246,8 +1246,8 @@ void __neigh_set_probe_once(struct neighbour *neigh)
  neigh->updated = jiffies;
  if (!(neigh->nud_state & NUD_FAILED))
   return;
- neigh->nud_state = NUD_PROBE;
- atomic_set(&neigh->probes, NEIGH_VAR(neigh->parms, UCAST_PROBES));
+ neigh->nud_state = NUD_INCOMPLETE;
+ atomic_set(&neigh->probes, neigh_max_probes(neigh));
  neigh_add_timer(neigh,
    jiffies + NEIGH_VAR(neigh->parms, RETRANS_TIME));
 }
-- 
1.7.1